src/compressed_tensors/utils/offload.py

```diff
@@ -49,68 +109,296 @@ def get_execution_device(module: Module) -> torch.device:
+    module.register_parameter(name, parameter)
+
+    if has_offloaded_params(module):
+        update_offload_parameter(module, name, parameter.data)
+        set_module_tensor_to_device(module, name, "meta")
+
+
+def update_offload_parameter(
+    module: torch.nn.Module,
+    name: str,
+    data: Optional[torch.Tensor],
+):
+    """
+    Update the data of an existing parameter and its offload dict. Supports both
+    parameters of offloaded modules and non-offloaded modules
+
+    :param module: module containing the parameter to update
+    :param name: name of module parameter to update
+    :param data: tensor to update parameter with
+    """
+    param = getattr(module, name)
+    data = data.to(param.dtype)
+
+    # copy data into onloaded parameter if applicable
+    if param.device != "meta":
+        param.data.copy_(data)
+
+    # update offload dict
+    if has_offloaded_params(module):
+        weights_map = module._hf_hook.weights_map
+        offload_to_weights_map(weights_map, name, data)
+
+
+def delete_offload_parameter(module: torch.nn.Module, name: str):
+    """
+    Delete a parameter from a module which may be offloaded
+
+    :param module: maybe offloaded module
+    :param name: name of parameter being deleted
+    """
+    delattr(module, name)
+
+    if has_offloaded_params(module):
+        weights_map = module._hf_hook.weights_map
+
+        # for upstreaming, better to add write capabilities to weight map classes first
+        if isinstance(weights_map, PrefixedDataset):
+            dataset = weights_map.dataset
+            prefix = weights_map.prefix
+            if dataset is not None:
+                del dataset[f"{prefix}{name}"]
 
-    parameter = getattr(module, param_name, None)
-    if parameter is None:
-        raise ValueError("Attempted to update uninitialized parameter")
+        elif isinstance(weights_map, dict):
+            del weights_map[name]
 
-    dtype = parameter.dtype
-    parameter.data = new_param_data.to(device).to(dtype)
+        elif isinstance(weights_map, OffloadedWeightsLoader):
+            raise NotImplementedError()
 
-    if offloaded:
-        prefix_dict = module._hf_hook.weights_map.dataset
-        prefix = module._hf_hook.weights_map.prefix
-        prefix_dict[f"{prefix}{param_name}"] = new_param_data.to(offload_device).to(
-            dtype
+        elif weights_map is not None:
+            raise NotImplementedError(
+                f"Cannot delete parameter from weights_map of type {type(weights_map)}"
+            )
+
+
+@check_accelerate(fallback=contextlib.nullcontext())
+@contextlib.contextmanager
+def disable_hf_hook(module: torch.nn.Module, recurse: bool = False):
+    hooks = {}
+
+    def collect_hooks(module):
+        nonlocal hooks
+        if hasattr(module, "_hf_hook"):
+            hooks[module] = module._hf_hook
+            remove_hook_from_module(module)
+
+        for submodule in module.children():
+            collect_hooks(submodule)
```

Comment:
can I do module.apply() here instead of looping the children?

Reply:
Yep!